content/en/blog/_posts/2022-10-27-live-and-let-live-with-kluctl-and-ssa.md

```diff
@@ -0,0 +1,172 @@
+which was still in beta at that time. Experimenting with it via
+`kubectl apply --server-side` revealed immediately that the true power of
+Server-side apply can not be easily leveraged by shelling out to kubectl.
+
+The way Server-side apply is implemented in kubectl does not allow enough
+control about conflict resolution as it can only switch between 
+"not force-applying anything and erroring out" and "force-apply everything
+without showing any mercy!".
+
+The API documentation however made it clear that Server-side apply is able to
+control conflict resolution on field level, simply by choosing which fields
+to include and which fields to omit from the supplied object.
+
+## Moving away from kubectl
+
+This meant that Kluctl had to move away from shelling out to kubectl first. Only
+after that was done, I would have been able to properly implement Server-side
+apply with powerful conflict resolution.
+
+To achieve this, I first implemented access to the target clusters via a
+Kubernetes client library. This had the nice side effect of dramatically
+speeding up Kluctl as well. It also improved the security and usability of
+Kluctl by ensuring that a running Kluctl command could not be messed around
+with by externally modifying the kubeconfig while it was running.
+
+## Implementing Server-side apply
+
+After switching to a Kubernetes client library, leveraging Server-side apply
+was easy. Kluctl just has to send each manifest to the apiserver and indicate
+that it wants to perform a Server-side apply operation. The apiserver will then
+either respond with an OK response or with a CONFLICT response.
+
+In case of a CONFLICT response, the apiserver also sends a status object with
+details about the conflicts. Kluctl can then use these details to figure out
+which fields are in conflict and which actors (field managers) have taken
+ownership of the conflicted fields.
+
+Then, for each field, Kluctl will decide if the conflict should be ignored or
+if it should be force-applied. If any field needs to be force-applied, Kluctl
+will retry the apply operation with the ignored fields omitted and the `force`
+flag being set on the API call.
+
+In case a conflict is ignored, Kluctl will issue a warning to the user so that
+the user can properly react (or ignore it forever...).
+
+That's basically it. That is all that is required to leverage Server-side apply.
+Big thanks and thumbs-up to the Kubernetes developers who made this possible!
+
+## Deciding which conflicts to ignore and which to force-apply
+
+Kluctl has a few simple rules to figure out if a conflict should be ignored
+or force-applied.
+
+It first checks the field's actor (the field manager) against a list of known
+tools that are well known to be used to perform manual modifications. These
+are for example kubectl and k9s. Any modifications performed with these tools
+are considered "temporary" and will be overwritten by Kluctl.
```

Comment:
```suggestion
Maybe mention that this has an implication if you combine Kluctl with a script that wraps
kubectl: Kluctl's heuristic will assume that the modifications by the script were temporary
/ ephemeral.
```

Reply:
I assume this was not meant as a code suggestion but as a comment? I'm not sure I understand tbh. Usually, when Kluctl is in use, there is no need to perform kubectl calls before or after so this situation should not arise. What I wrote here was meant for cases where a modification is done outside of the deployment process.

Reply:
Oops, yes.